e2e/tests/upgrade-test.rs

```diff
@@ -1,36 +1,438 @@
+    install_promscale_ext(&mut baseline_client, &to_version);
+
+    // load test data if configured to
+    if with_data {
+        load_data(&baseline_container);
+    }
+
+    // snapshot the database
+    let baseline_snapshot: String = snapshot_db(
+        baseline_container.id(),
+        "db",
+        "postgres",
+        working_dir.join("baseline-snapshot.txt").as_path(),
+    );
+
+    // shut it down
+    baseline_client
+        .close()
+        .expect("failed to close database connection");
+    baseline_container.stop();
+
+    /**********************************************************************************************/
+    // UPGRADE FROM
+    info!("UPGRADE FROM");
+    let from_blueprint = PostgresContainerBlueprint::new()
+        .with_image_uri(from_image_uri)
+        .with_volume(script_dir, "/scripts")
+        .with_volume(data_dir, volume_map_path)
+        .with_env_var("PGDATA", "/var/lib/postgresql/data")
+        .with_db("db")
+        .with_user("postgres");
+    info!("run container");
+    let from_container = from_blueprint.run();
+    info!("connect");
+    let mut from_client = connect(&from_blueprint, &from_container);
+
+    // install the timescaledb extension at the default version
+    info!("install ext");
+    let from_timescaledb_version = install_timescaledb_ext(&mut from_client);
+    drop_toolkit_extension(&mut from_client);
+    info!("from timescaledb version: {}", &from_timescaledb_version);
+    info!("to timescaledb version: {}", &to_timescaledb_version);
+
+    // if the "from" version is greater than the "to" version, then we have problems!
+    assert!(from_timescaledb_version <= to_timescaledb_version);
+
+    // install the promscale extension at the "from_version"
+    install_promscale_ext(&mut from_client, &from_version);
+
+    // load test data if configured to
+    if with_data {
+        load_data(&from_container);
+    }
+
+    // checkpoint
+    from_client
+        .execute("checkpoint", &[])
+        .expect("failed to checkpoint");
+
+    // shut it down
+    from_client
+        .close()
+        .expect("failed to close database connection");
+    from_container.stop();
+
+    /**********************************************************************************************/
+    // UPGRADE TO
+    info!("UPGRADE TO");
+    let to_blueprint = PostgresContainerBlueprint::new()
+        .with_image_uri(to_image_uri)
+        .with_volume(script_dir, "/scripts")
+        .with_volume(data_dir, volume_map_path)
+        .with_env_var("PGDATA", "/var/lib/postgresql/data")
+        .with_db("db")
+        .with_user("postgres");
+    let to_container = to_blueprint.run();
+    info!("to_container id: {}", to_container.id());
+
+    // upgrade the timescaledb extension if we need to
+    if from_timescaledb_version != to_timescaledb_version {
+        info!(
+            "upgrading from timescaledb {} to {}",
+            from_timescaledb_version, to_timescaledb_version
```

Comment:
Why does it have to happen via psql?

Reply:
I'm actually not sure if it's necessary. The timescaledb upgrade is quite finnicky though, the upgrade command has to be the only command executed on the connection, otherwise weird stuff happens. I will fiddle with this a bit and see.

Reply:
The timescaledb docs say that you need to pass the -X flag to psql for the upgrade to work. This prevents the extension from being loaded into the session. I tried it without psql first and it failed, complaining about the extension having been loaded already